tests/unit/modules/test_boto_cfn.py

```diff
@@ -0,0 +1,163 @@
+
+    def setUp(self):
+        super().setUp()
+        boto_cfn.__init__(self.opts)
+        del self.opts
+
+        self.conn_parameters = {
+            "region": self.region,
+            "key": self.access_key,
+            "keyid": self.secret_key,
+            "profile": {},
+        }
+
+        self.mock = mock_cloudformation_deprecated()
+        self.mock.start()
+        self.conn = boto.connect_cloudformation()
+
+    def tearDown(self):
+        self.mock.stop()
+
+    def test_stack_exists_that_does_not_exist(self):
+        name = _random_stack_name()
+        self.assertFalse(boto_cfn.exists(name, **self.conn_parameters))
+
+    def test_stack_exists_that_does_exist(self):
+        name = _random_stack_name()
+        resp = self.conn.create_stack(name, self.template_body)
+        self.assertTrue(boto_cfn.exists(name, **self.conn_parameters))
+
+    def test_stack_describe_returns_correct_json(self):
+        name = _random_stack_name()
+        self.conn.create_stack(name, self.template_body)
+        resp = self.conn.describe_stacks(name)
+
+        # Note: this was lifted almost 100% from the boto_cfn code.  The format
+        # of this data has changed dramatically between boto2 and boto3, and we
+        # need to preserve that format if we are to maintain compatibility.
+        stack = resp[0]
+        keys = (
+            "stack_id",
+            "description",
+            "stack_status",
+            "stack_status_reason",
+            "tags",
+        )
+        data = {k: getattr(stack, k) for k in keys if hasattr(stack, k)}
+        o = getattr(stack, "outputs")
+        p = getattr(stack, "parameters")
+        outputs = {}
+        parameters = {}
+        for i in o:
+            outputs[i.key] = i.value
+        data["outputs"] = outputs
+        for j in p:
+            parameters[j.key] = j.value
+        data["parameters"] = parameters
+        stack = {"stack": data}
+
+        self.assertEqual(boto_cfn.describe(name, **self.conn_parameters), stack)
+
+    def test_passing_templatebody_and_template_url_raises_exception(self):
+        name = _random_stack_name()
+        with self.assertRaises(SaltInvocationError):
+            boto_cfn.create(
+                name, self.template_body, "s3://anywhere", **self.conn_parameters
+            )
+
+    def test_stack_create_new_stack_returns_true(self):
+        name = _random_stack_name()
+        self.assertTrue(
+            boto_cfn.create(name, self.template_body, **self.conn_parameters)
+        )
+
+    def test_stack_create_existing_stack_returns_true(self):
+        name = _random_stack_name()
+        resp = self.conn.create_stack(name, self.template_body)
+        self.assertTrue(
+            boto_cfn.create(name, self.template_body, **self.conn_parameters)
+        )
+
+    def test_delete_none_existant_stack_returns_success(self):
+        name = _random_stack_name()
+        self.assertTrue(boto_cfn.delete(name, **self.conn_parameters))
```

Comment:
Should we be failing if we are trying to delete something that does not exist? If so, we can try to handle this outside of this PR. I can create an issue if this is the case.

Reply:
I am not 100% certain here.  Some of the other boto modules return true when deleting something that does not exist (or creating something that already exists), so I was mostly just mimic'ing that behavior.

Sitting here looking through some of the other modules and it unfortunately looks like a mixed bag.

Reply:
I was just looking at `boto_vpc` and noticed that it returns does return `False` when attempting to delete something that does not exist, but it returns it as part of a dictionary such that it can return a status _and_ a message. Should this be the preferred approach for these modules?

Reply:
It looks like `boto_iam` is the primary source which returns true if the target resource already exists:
```
def create_instance_profile(name, region=None, key=None, keyid=None, profile=None):
    """
    Create an instance profile.

    CLI Example:

    .. code-block:: bash

        salt myminion boto_iam.create_instance_profile myiprofile
    """
    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)

    if instance_profile_exists(name, region, key, keyid, profile):
        return True
    try:
        # This call returns an instance profile if successful and an exception
        # if not. It's annoying.
        conn.create_instance_profile(name)
        log.info("Created %s instance profile.", name)
    except boto.exception.BotoServerError as e:
        log.debug(e)
        log.error("Failed to create %s instance profile.", name)
        return False
    return True
```

IMHO I prefer the approach that boto_vpc takes with returning false on these attempts, but also returning an error message with it.

Reply:
Yeah I agree with your approach as well, lthough this would need to be handled with great care since the state modules might not be accounting for this change. ping @waynew thoughts? Do you agree or should we keep as is?

Reply:
I think I don't know enough to make a *specific* recommendation.

Here's what I /do/ know:

- Some things in Salt return True when a thing is already gone.
- Some things in Salt return True (with `'changes':`) when something is already gone.
- Some things in Salt return False when they can't do something they wanted to do.

As I understand it, the question is basically one of intent. If the "universe" is in the desired state, then the result is True, regardless of how we got there. If the universe is *not* in the desired state, then the result should be False.

From what little I understand at the moment it looks like the /desired/ state is that the thing is deleted and gone. If that's the case, then whether it was deleted with fire, or we tried to delete it and it was already gone, the universe looks like we desire it, and it should return True.


But I could be wrong! 🙃  (but hopefully the reasoning behind it is clear enough - let me know if not!)

Reply:
@waynew That seems to align with the code in question.  In particular some code has an explicit test to return true if the desired state already exists (since the boto layer might return an error otherwise), and some code does not have a test (though still returns true because the AWS boto layer does not error). For example, the AWS `boto.create_stack()` call returns true even if the stack already exists, where the similar AWS boto IAM call will throw an error if you true to create a user that already exists (so `boto_iam.create_user()` performs extra tests).

While digging deeper into this, I realized that nearly all of the boto_cfn code simply returns a `str` on error (taken from the BotoServerError), which means it wasn't possible to test for an error based on truthyness. Further more, `boto_cfn.delete()` returned the boto response dictionary on success (i.e. `return {'DeleteStackResponse': {'DeleteStackResult': {}}}`), and a `str` on error.